core/drivers/firewall/stm32_risaf.c

```diff
@@ -0,0 +1,770 @@
+
+static vaddr_t risaf_base(struct stm32_risaf_instance *risaf)
+{
+	return io_pa_or_va_secure(&risaf->pdata.base, 1);
+}
+
+void stm32_risaf_clear_illegal_access_flags(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
+
+		if (clk_enable(risaf->pdata.clock))
+			panic("Can't enable RISAF clock");
+
+		if (!io_read32(base + _RISAF_IASR)) {
+			clk_disable(risaf->pdata.clock);
+			continue;
+		}
+
+		io_write32(base + _RISAF_IACR, _RISAF_IACR_CAEF |
+			   _RISAF_IACR_IAEF0 | _RISAF_IACR_IAEF1);
+
+		clk_disable(risaf->pdata.clock);
+	}
+}
+
+#ifdef CFG_TEE_CORE_DEBUG
+void stm32_risaf_dump_erroneous_data(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
+
+		if (clk_enable(risaf->pdata.clock))
+			panic("Can't enable RISAF clock");
+
+		/* Check if faulty address on this RISAF */
+		if (!io_read32(base + _RISAF_IASR)) {
+			clk_disable(risaf->pdata.clock);
+			continue;
+		}
+
+		EMSG("\n\nDUMPING DATA FOR %s\n\n", risaf->pdata.risaf_name);
+		EMSG("=====================================================");
+		EMSG("Status register (IAESR0): %#"PRIx32,
+		     io_read32(base + _RISAF_IAESR0));
+
+		/* Reserved if dual port feature not available */
+		if (io_read32(base + _RISAF_IAESR1))
+			EMSG("Status register Dual Port (IAESR1) %#"PRIx32,
+			     io_read32(base + _RISAF_IAESR1));
+
+		EMSG("-----------------------------------------------------");
+		if (virt_to_phys((void *)base) == RISAF4_BASE) {
+			EMSG("Faulty address (IADDR0): %#"PRIxPA,
+			     risaf->pdata.mem_base +
+			     (paddr_t)io_read32(base + _RISAF_IADDR0));
+
+			/* Reserved if dual port feature not available */
+			if (io_read32(base + _RISAF_IADDR1))
+				EMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
+				     risaf->pdata.mem_base +
+				     (paddr_t)io_read32(base + _RISAF_IADDR1));
+		} else {
+			EMSG("Faulty address (IADDR0): %#"PRIxPA,
+			     (paddr_t)io_read32(base + _RISAF_IADDR0));
+
+			/* Reserved if dual port feature not available */
+			if (io_read32(base + _RISAF_IADDR1))
+				EMSG("Dual port faulty address (IADDR1): %#"PRIxPA,
+				     (paddr_t)io_read32(base + _RISAF_IADDR1));
+		}
+
+		EMSG("=====================================================\n");
+
+		clk_disable(risaf->pdata.clock);
+	};
+}
+#endif
+
+static __maybe_unused
+bool risaf_is_hw_encryption_enabled(struct stm32_risaf_instance *risaf)
+{
+	return (io_read32(risaf_base(risaf) + _RISAF_SR) &
+		_RISAF_SR_ENCDIS) != _RISAF_SR_ENCDIS;
+}
+
+static TEE_Result
+risaf_check_region_boundaries(struct stm32_risaf_instance *risaf,
+			      struct stm32_risaf_region *region)
+{
+	uintptr_t end_paddr = 0;
+
+	assert(region->len != 0U);
+
+	if (region->addr < risaf->pdata.mem_base) {
+		EMSG("RISAF %#"PRIxPTR": region start address lower than memory base",
```

Comment:
s/PRIxPTR/PRIxPA/   for `struct io_pa_va :: paddr_t pa`
Ditto at lines 300, 308, 331, 364, 370, 376, 398, 645, 711.